### PR TITLE
bpo-29447: tempfile: Add pathlike object support for dir argument

### PR DESCRIPTION
**Pull request by louisom**

### CODE REVIEW

File: Doc/library/tempfile.rst [hunk 1]
```diff
@@ -66,6 +66,11 @@ The module defines the following user-callable items:
 
       The :py:data:`os.O_TMPFILE` flag is now used if available.
 
+   .. versionchanged:: 3.7
```
Comment (brettcannon) [inline Doc/library/tempfile.rst]: So this isn't entirely true since you didn't need to change anything for support, so this really has been supported since Python 3.6.

File: Doc/library/tempfile.rst [hunk 2]
```diff
@@ -66,6 +66,11 @@ The module defines the following user-callable items:
...
+
+      The ``dir`` argument is now accept a :term:`path-like object`.
```
Comment (brettcannon) [inline Doc/library/tempfile.rst]: ``The *dir* argument now accepts a :term:`path-like object`.``

File: Lib/test/test_tempfile.py [hunk 3]
```diff
@@ -427,6 +450,14 @@ def test_choose_directory(self):
         finally:
             os.rmdir(dir)
 
+    def test_choose_pathlike_directory(self):
+        # _mkstemp_inner can create files in a user-selected pathlike directory
+        dir = pathlib.Path(tempfile.mkdtemp())
+        try:
```
Comment (brettcannon) [inline Lib/test/test_tempfile.py]: Instead of a try/finally, use [`unittest.TestCase.addCleanup()`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.addCleanup) (basically you shouldn't have to use any try/finally blocks in your test code thanks to that method).

File: Misc/NEWS [hunk 4]
```diff
@@ -371,6 +371,8 @@ Library
 - bpo-30068: _io._IOBase.readlines will check if it's closed first when
   hint is present.
 
+- bpo-29447: tempfile's dir argument is now accept path-like objects.
```
Comment (brettcannon) [inline Misc/NEWS]: This isn't accurate since the support was already there implicitly, so just drop the changes to this file or add a comment that tests for path-like objects in tempfile were added to the "Tests" section.

File: Lib/test/test_tempfile.py [hunk 5]
```diff
@@ -39,27 +40,46 @@
 class TestLowLevelInternals(unittest.TestCase):
     def test_infer_return_type_singles(self):
         self.assertIs(str, tempfile._infer_return_type(''))
+        self.assertIs(str, tempfile._infer_return_type(pathlib.Path('')))
```
Comment (brettcannon) [inline Lib/test/test_tempfile.py]: It wouldn't hurt to add a test for a path-like object that returns bytes (you will have to make a test object yourself).

File: Lib/test/test_tempfile.py [hunk 6]
```diff
@@ -636,6 +667,7 @@ def test_basic(self):
         self.do_create(pre="a", suf="b")
         self.do_create(pre="aa", suf=".txt")
         self.do_create(dir=".")
+        self.do_create(dir=pathlib.Path("."))
 
     def test_basic_with_bytes_names(self):
```
Comment (brettcannon) [inline Lib/test/test_tempfile.py]: Add a test with a path-like object that returns bytes.

**Review by brettcannon — CHANGES_REQUESTED** [review]
Don't forget to add yourself to Misc/ACKS.

File: Doc/library/tempfile.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/library/tempfile.rst]: I think no documentation changes is needed. Otherwise we should add "versionchanged" directives for every function that got an implicit support of path-like objects (this is perhaps many tens of functions). This just distracts the attention.

File: Lib/test/test_tempfile.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/test/test_tempfile.py]: There are test path-like classes in other test files. The code can be shared.

File: Misc/NEWS (hunk 4 above)
Comment (serhiy-storchaka) [reply Misc/NEWS]: Usually we don't add entries about changes to tests unless this is very large change or a change that affects the testing infrastructure (test.support, test.libregrtest).

File: Doc/library/tempfile.rst (hunk 2 above)
Comment (louisom) [reply Doc/library/tempfile.rst]: Revert the docs change for now.

File: Lib/test/test_tempfile.py (hunk 5 above)
Comment (louisom) [reply Lib/test/test_tempfile.py]: I couldn't find other place test path-like with bytes. Add some test here, and at low-level internals test

File: Misc/NEWS (hunk 4 above)
Comment (louisom) [reply Misc/NEWS]: Remove the NEWS entry.

File: Lib/test/test_tempfile.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/test/test_tempfile.py]: Search `def __fspath__` in tests.

File: Lib/tempfile.py [hunk 7]
```diff
@@ -98,7 +98,7 @@ def _infer_return_type(*args):
     """Look at the type of all args and divine their implied return type."""
     return_type = None
     for arg in args:
-        if arg is None:
+        if arg is None or isinstance(arg, _os.PathLike):
```
Comment (brettcannon) [inline Lib/tempfile.py]: Why is this necessary? Can't you simply call `_os.fspath(arg)` just before `if isinstance(arg, bytes)`?

File: Lib/tempfile.py [hunk 8]
```diff
@@ -130,6 +130,11 @@ def _sanitize_params(prefix, suffix, dir):
             dir = gettempdir()
         else:
             dir = gettempdirb()
+    if isinstance(dir, _os.PathLike):
```
Comment (brettcannon) [inline Lib/tempfile.py]: Why is the encoding call suddenly necessary when it wasn't needed for `dir` previously?

**Review by brettcannon — CHANGES_REQUESTED** [review]
If those changes you made to tempfile are necessary then the doc changes you reverted will need to come back as this is no longer a patch to just add some tests.

File: Lib/tempfile.py (hunk 7 above)
Comment (serhiy-storchaka) [reply Lib/tempfile.py]: I think `dir = _os.fspath(dir)` should be called before calling `_infer_return_type()`. It would be not correct to call `_os.fspath()` for non-path arguments.

Comment (brettcannon) [reply Lib/tempfile.py]: @serhiy-storchaka sounds reasonable to me.

Comment (louisom) [reply Lib/tempfile.py]: We can't do `dir = _os.fspath(dir)` too fast, otherwise path-like objects will only be treat as an `str`, not `str` or `bytes`.

File: Lib/tempfile.py (hunk 8 above)
Comment (louisom) [reply Lib/tempfile.py]: We should fist checkout the output type, to respect other arguments (prefix and suffix), then decide how to convert dir when it is a path-like objects.

File: Lib/tempfile.py (hunk 7 above)
Comment (serhiy-storchaka) [reply Lib/tempfile.py]: `os.fspath()` returns `str` or `bytes`.

Comment (louisom) [reply Lib/tempfile.py]: ah, I get it. Is my misunderstand about path-like objects, I thought path-like object should return string only. because stdlib implement `__fspath__` with return `str` only. And there should be some 3rd-party class implement `__fspath__` with return `bytes`.

Comment (serhiy-storchaka) [reply Lib/tempfile.py]: For example `DirEntry.__fspath__` can return `bytes`.

File: Lib/tempfile.py [hunk 9]
```diff
@@ -100,6 +100,8 @@ def _infer_return_type(*args):
     for arg in args:
         if arg is None:
             continue
+        elif isinstance(arg, _os.PathLike):
```
Comment (serhiy-storchaka) [inline Lib/tempfile.py:103]: Why this explicit test is needed? Aren't high-level function fail in any case when pass path-like object as prefix or suffix?

Comment (louisom) [reply Lib/tempfile.py:103]: Should we allow path-like objects for prefix and suffix? it will failed when path-like objects concat with string:
`file = _os.path.join(dir, pre + name + suf)`

Comment (brettcannon) [reply Lib/tempfile.py:103]: @lulouie No, path-like objects are only useful for things that represent paths. Both `prefix` and `suffix` are not paths themselves.

File: Lib/test/test_tempfile.py [hunk 10]
```diff
@@ -32,6 +32,14 @@
 else:
     TEST_FILES = 100
 
+
+class _PathLikeObj(object):
```
Comment (serhiy-storchaka) [inline Lib/test/test_tempfile.py:36]: Use `test.support.FakePath`.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Path-like objects shouldn't be accepted as *prefix* or *suffix*. There should be tests for this.

**Comment by csabella** [comment]
As this pull request is from an 'unknown repository', I'm going to close it.  A new pull request can be opened with this change from a valid repository.
